cirq-core/cirq/transformers/dynamical_decoupling.py

```diff
@@ -0,0 +1,161 @@
+        raise ValueError(
+            "Invalid dynamical decoupling sequence. Expect sequence production equals identity"
+            f" up to a global phase, got {product}.".replace('\n', ' ')
+        )
+
+
+@value.value_equality
+class DynamicalDecouplingModel:
+    """Dynamical decoupling model that generates dynamical decoupling operation sequences."""
+
+    def __init__(
+        self,
+        schema: Optional[_DynamicalDecouplingSchema] = None,
+        base_dd_sequence: Optional[list['cirq.Gate']] = None,
+    ):
+        if not schema and not base_dd_sequence:
+            raise ValueError(
+                'Specify either schema or base_dd_sequence to construct a valid'
+                ' DynamicalDecouplingModel.'
+            )
+        self.schema = schema
+        self.base_dd_sequence = base_dd_sequence
+        if base_dd_sequence:
+            _validate_dd_sequence(base_dd_sequence)
+
+    def generate_dd_sequence(self, num_idle_moments: int = 2) -> list['cirq.Gate']:
+        """Returns the longest possible dynamical decoupling sequence."""
+        if num_idle_moments <= 0:
+            return []
+        if self.schema:
+            dd_sequence = _generate_dd_sequence_from_schema(self.schema, num_idle_moments)
+        elif self.base_dd_sequence:
+            dd_sequence = _repeat_sequence(self.base_dd_sequence, num_idle_moments)
+        return dd_sequence
+
+    @classmethod
+    def from_schema(cls, schema: str):
+        """Create dynamical decoupling model according to a given schema."""
+        if not schema in _DynamicalDecouplingSchema.__members__:
+            raise ValueError("Invalid schema name.")
+        return cls(schema=_DynamicalDecouplingSchema[schema])
+
+    @classmethod
+    def from_base_dd_sequence(cls, base_dd_sequence: list['cirq.Gate']):
+        """Create dynamical decoupling model according to a base sequence."""
+        return cls(base_dd_sequence=base_dd_sequence)
+
+    def _json_dict_(self) -> Dict[str, Any]:
+        d: Dict[str, Any] = {}
+        if self.schema:
+            d['schema'] = self.schema.name
+        if self.base_dd_sequence:
+            d['base_dd_sequence'] = self.base_dd_sequence
+        return d
+
+    @classmethod
+    def _from_json_dict_(cls, schema=None, base_dd_sequence=None, **kwargs):
+        if schema:
+            return cls(schema=_DynamicalDecouplingSchema[schema])
+        if base_dd_sequence:
+            return cls(base_dd_sequence=base_dd_sequence)
+
+    def _value_equality_values_(self) -> Any:
+        return self.schema, self.base_dd_sequence
+
+
+@transformer_api.transformer
+def add_dynamical_decoupling(
+    circuit: 'cirq.AbstractCircuit',
+    *,
+    context: Optional['cirq.TransformerContext'] = None,
+    dd_model: DynamicalDecouplingModel = DynamicalDecouplingModel.from_schema("X_XINV"),
+) -> 'cirq.Circuit':
+    """Add dynamical decoupling gate operations to a given circuit.
+
```

Comment:
does this preserve the moment structure? consider adding this to the docstring.

Reply:
It preserves the moment structure.

Updated docstring.